src/Features/Core/Portable/PullMemberUp/MembersPuller.cs

```diff
@@ -0,0 +1,408 @@
+        internal async Task<Solution> PullMembersUpAsync(
+            Document document,
+            PullMembersUpAnalysisResult result,
+            CancellationToken cancellationToken)
+        {
+            if (result.Destination.TypeKind == TypeKind.Interface)
+            {
+                return await PullMembersIntoInterfaceAsync(document, result, cancellationToken).ConfigureAwait(false);
+            }
+            else if (result.Destination.TypeKind == TypeKind.Class)
+            {
+                return await PullMembersIntoClassAsync(document, result, cancellationToken).ConfigureAwait(false);
+            }
+            else
+            {
+                throw ExceptionUtilities.UnexpectedValue(result.Destination);
+            }
+        }
+
+        private IMethodSymbol FilterGetterOrSetter(IMethodSymbol getterOrSetter)
+        {
+            // We are pulling a public property, it could have a public getter/setter but 
+            // the other getter/setter is not.
+            // In this scenario, only the public getter/setter
+            // will be add to the destination interface.
+            return getterOrSetter?.DeclaredAccessibility == Accessibility.Public ? getterOrSetter : null;
+        }
+
+        private IMethodSymbol CreatePublicGetterAndSetter(IMethodSymbol getterOrSetter)
+        {
+            // Create a public getter/setter since user is trying to pull a non-public property to an interface.
+            // If getterOrSetter is null, it means this property doesn't have a getter/setter, so just don't generate it.
+            return getterOrSetter == null
+                ? getterOrSetter
+                : CodeGenerationSymbolFactory.CreateMethodSymbol(getterOrSetter, accessibility: Accessibility.Public);
+        }
+
+        private async Task<Solution> PullMembersIntoInterfaceAsync(
+            Document document,
+            PullMembersUpAnalysisResult result,
+            CancellationToken cancellationToken)
+        {
+            var solution = document.Project.Solution;
+            var solutionEditor = new SolutionEditor(solution);
+            var codeGenerationService = document.Project.LanguageServices.GetRequiredService<ICodeGenerationService>();
+            var destinationSyntaxNode = await codeGenerationService.FindMostRelevantNameSpaceOrTypeDeclarationAsync(
+                document.Project.Solution, result.Destination, default, cancellationToken).ConfigureAwait(false);
+            var (editorMap, syntaxMap) = await InitializeEditorMapsAndSyntaxMapAsync(result, solution, solutionEditor, destinationSyntaxNode, cancellationToken).ConfigureAwait(false);
+            var symbolsToPullUp = result.MemberAnalysisResults.
+                SelectAsArray(analysisResult =>
+                {
+                    if (analysisResult.Member is IPropertySymbol propertySymbol)
+                    {
+                        if (analysisResult.ChangeOriginalToPublic)
```

Comment:
what i don't quite undertand is why this is only for properties.  i.e. you say:

> We are pulling a non-public property, change ... itself to be public.

Why does this not apply to anything you are pulling up?  Why do you ony need to do this for properties?

Reply:
So other members can be directly put into ICodeGenerationService and set the target as interface, CodeGenerationService then take care of everything and generate the declarations. 
But for property it is not true, as for non-public property, like private int Bar {get; set;}, codeGenerationSerivce will generate a property like this 'Bar {}' (It filter out all private getter/setter and leave an empty {} ) So that why I try to generate a public getter/setter here for codeGeneratationService. 

I can put comments here to explain it.